tidy3d/components/device/simulation.py

```diff
@@ -318,18 +356,80 @@ def check_medium_heat_spec(cls, values):
-                "No solid materials ('SolidSpec') are detected in heat simulation. Solution domain is empty."
-            )
+        simulation_types = cls._check_simulation_types(values=values)
+
+        for sim_type in simulation_types:
+            if sim_type == DeviceSimulationType.HEAT:
+                if len(failed_solid_idx) > 0:
+                    raise SetupError(
+                        "No solid materials ('SolidSpec') are detected in heat simulation. Solution domain is empty."
+                    )
+            elif sim_type == DeviceSimulationType.CONDUCTION:
+                if len(failed_elect_idx) > 0:
+                    raise SetupError(
+                        "No conducting materials ('ConductorSpec') are detected in conduction simulation. Solution domain is empty."
+                    )
+
+        return values
+
+    @staticmethod
+    def _check_simulation_types(
+        values: Dict,
+        HeatBCTypes=HeatBCTypes,
+        ElectricBCTypes=ElectricBCTypes,
+        HeatSourceTypes=HeatSourceTypes,
+    ) -> list[DeviceSimulationType]:
+        """Given model dictionary ``values``, check the type of simulations to be run
+        based on BCs and sources.
+        """
+        # NOTE: this function does not consider the source 'HeatFromElectricSource' by default
+
+        boundaries = values["boundary_spec"]
+        sources = values["sources"]
+
+        simulation_types = []
+
+        for boundary in boundaries:
+            if isinstance(boundary.condition, HeatBCTypes):
+                if DeviceSimulationType.HEAT not in simulation_types:
+                    simulation_types.append(DeviceSimulationType.HEAT)
+            if isinstance(boundary.condition, ElectricBCTypes):
+                if DeviceSimulationType.CONDUCTION not in simulation_types:
+                    simulation_types.append(DeviceSimulationType.CONDUCTION)
+
+        for source in sources:
+            if isinstance(source, HeatSourceTypes):
+                if DeviceSimulationType.HEAT not in simulation_types:
+                    simulation_types.append(DeviceSimulationType.HEAT)
+
+        return simulation_types
+
+    @pd.root_validator(skip_on_failure=True)
+    def check_coupling_source_can_be_applied(cls, values):
+        """Error if material doesn't have the right specifications"""
+
+        HeatSourceTypes_noCoupling = (UniformHeatSource, HeatSource)
+
+        simulation_types = cls._check_simulation_types(
+            values, HeatSourceTypes=HeatSourceTypes_noCoupling
+        )
+
+        sources = values["sources"]
+
+        for source in sources:
+            if isinstance(source, HeatFromElectricSource) and len(simulation_types) < 2:
+                raise SetupError(
+                    "Using 'HeatFromElectricSource' requires the definition of both CONDUCTION and "
+                    f"HEAT simulations. Your simulation setup contains only conditions of type {simulation_types[0].name}"
```

Comment:
can `simulation_types` be empty?

Reply:
it shouldn't. We have another validator that would raise an error if no BCs are specified.